### ISSUE TITLE
macOS (and arm64) Support - Feature Request

### ISSUE THREAD

**Issue by ilgrandeanonimo** [Numelon-Softworks/ThreatenGL#11]
Today i tried to use this mod on my M1 Mac but the game crash because it requires the ARM64 version
```java
[22:58:41] [Render thread/ERROR]:
 Unhandled game exception
java.lang.UnsatisfiedLinkError: dlopen(/Users/{COMPUTER_USERNAME}/Library/Application Support/com.modrinth.theseus/meta/natives/1.20.6-0.15.11/jna15495653446642276667.tmp, 0x0009): tried: '/Users/{COMPUTER_USERNAME}/Library/Application Support/com.modrinth.theseus/meta/natives/1.20.6-0.15.11/jna15495653446642276667.tmp' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e' or 'arm64')), '/System/Volumes/Preboot/Cryptexes/OS/Users/{COMPUTER_USERNAME}/Library/Application Support/com.modrinth.theseus/meta/natives/1.20.6-0.15.11/jna15495653446642276667.tmp' (no such file), '/Users/{COMPUTER_USERNAME}/Library/Application Support/com.modrinth.theseus/meta/natives/1.20.6-0.15.11/jna15495653446642276667.tmp' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e' or 'arm64'))
	at com.sun.jna.Native.open(Native Method) ~[jna-5.14.0.jar:?]
	at com.sun.jna.NativeLibrary.loadLibrary(NativeLibrary.java:300) ~[jna-5.14.0.jar:?]
	at com.sun.jna.NativeLibrary.getInstance(NativeLibrary.java:481) ~[jna-5.14.0.jar:?]
	at com.sun.jna.Library$Handler.<init>(Library.java:197) ~[jna-5.14.0.jar:?]
	at com.sun.jna.Native.loadLibrary(Native.java:668) ~[jna-5.14.0.jar:?]
	at com.sun.jna.Native.loadLibrary(Native.java:652) ~[jna-5.14.0.jar:?]
	at club.minnced.discord.rpc.DiscordRPC.<clinit>(DiscordRPC.java:42) ~[com_github_minndevelopment_java--rpc-v2.0.2-15f21a422df796c6.jar:?]
	at me.splitque.presencium.discord.<clinit>(discord.java:8) ~[presencium-2.0.jar:?]
	at me.splitque.presencium.main.lambda$onInitialize$0(main.java:15) ~[presencium-2.0.jar:?]
	at net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientLifecycleEvents.lambda$static$0(ClientLifecycleEvents.java:35) ~[fabric-lifecycle-events-v1-2.3.4+c5fc38b3b0-b28198e202c52719.jar:?]
	at net.minecraft.class_310.handler$zll000$fabric-lifecycle-events-v1$onStart(class_310.java:7258) ~[client-intermediary.jar:?]
	at net.minecraft.class_310.method_1514(class_310.java:870) ~[client-intermediary.jar:?]
	at net.minecraft.client.main.Main.main(Main.java:265) [client-intermediary.jar:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:470) [fabric-loader-0.15.11.jar:?]
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74) [fabric-loader-0.15.11.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23) [fabric-loader-0.15.11.jar:?]
```

**Comment by ilgrandeanonimo** [comment]
I just realized the issue is related to another mod. Sry for bothering you.

**Comment by Richy-Z** [comment]
I wanted to reopen this issue to inform future users that this mod isn’t fully supported on macOS. While it does work on M series Macs due to the platform-independent nature of JAR files (which can run on almost anything, even Samsung Smart Fridges), the mod performs a platform check and only supports **up to OpenGL 4.3 on macOS**. This limitation is due to Apple’s decision to drop OpenGL in favour of the Metal API (similar to DirectX or Vulkan) back in the Mojave era, leaving OpenGL support capped at version 4.3. This is not a restriction imposed by me, Minecraft, OpenGL, GLFW, or any other entity, but Apple.

**Comment by ilgrandeanonimo** [comment]
Maybe instead of OpenGL, on Mac computers you could rely on Metal APIs

**Comment by Richy-Z** [comment]
That's an intriguing concept that I looked into a while ago. Unfortunately, it is impossible because there are no Metal Framework bindings for Java. The closest thing I could find was [Project Lanai](https://wiki.openjdk.org/display/lanai/Main) and [JEP 382](https://openjdk.org/jeps/382), but both are absolutely useless in our case because they are simply the implementation of Metal to replace OpenGL in 2D Java Applications. Additionally, it is incredibly limited and the Project Lanai wiki makes it hard to understand *what it is* - it's as if they're obscuring what the project actually entails. They could have simply written, "Project Lanai is the attempt to implement Metal to replace OpenGL for 2D Java Applications on macOS," but instead, it is a disorganized mess of information on an obscure, outdated wiki.

I'd love to see something like Vulkan mod except Metal for macOS one day, if bindings become available. The performance impact for M series Macs would be immense.

TL;DR: Can't make a Metal Mod for Minecraft because of no available bindings for Java. Looked at Project Lanai but that's only for 2D.

**Comment by ilgrandeanonimo** [comment]
Maybe is possible to use Metal APIs in Java by mixing Java with Swift https://github.com/SwiftJava or C++

**Comment by ilgrandeanonimo** [comment]
Swift team has just announced a new package for Java interoperability https://github.com/swiftlang/swift-java.

**Comment by Richy-Z** [comment]
That seems like an interesting thing. I'll see how difficult it is when I have some spare time.

**Comment by Richy-Z** [comment]
I'm reopening this issue as I'm partially interested in this concept after June 20.

**Comment by ilgrandeanonimo** [comment]
When I opened the issue, I mistakenly confused an error from discordrpc with threatengl. Minecraft already use Metal with a translation layer. I'm not sure that using Metal directly would improve performance.

**Comment by Richy-Z** [comment]
Hmm.. that doesnt seem right. Are you sure about this?
Minecraft still definitely uses OpenGL at its core, so I'm going to assume that youre talking about macOS itself translating OpenGL calls to Metal. If we're on the same page, then I would assume that this translation only happens on Apple Silicon Macs because M-series processors do not natively support OpenGL. But for the Macs that have been left behind with Intel (& their integrated graphics or AMD, etc), I'm not entirely sure what translation layer youre referring to.

**Comment by Richy-Z** [comment]
Actually, upon some further digging, there doesn't seem to be a translation layer for OpenGL on M-series Macs either.
I must also correct myself that OpenGL ***is*** supported on Apple Silicon, but it is more limited than Intel macs

**Comment by Richy-Z** [comment]
In this case, I suppose that the success of the Vulkan mod demonstrates why manual optimisation with native or better graphics APIs matter. Feel free to correct me/provide more info though haha 😅

**Comment by ilgrandeanonimo** [comment]
It's automatic on M-series mac. When I play Minecraft I have Metal 4.1 (Metal + OpenGL version)
